rmw_connext_shared_cpp/src/node_names.cpp

```diff
@@ -90,44 +104,100 @@ get_node_names(
       auto data = static_cast<unsigned char *>(pbtd.user_data.value.get_contiguous_buffer());
       std::vector<uint8_t> kv(data, data + pbtd.user_data.value.length());
+
       auto map = rmw::impl::cpp::parse_key_value(kv);
+
       auto name_found = map.find("name");
       auto ns_found = map.find("namespace");
 
       if (name_found != map.end()) {
         name = std::string(name_found->second.begin(), name_found->second.end());
       }
-      if (name_found != map.end()) {
+
+      if (ns_found != map.end()) {
         namespace_ = std::string(ns_found->second.begin(), ns_found->second.end());
       }
+      
       if (name.empty()) {
         // use participant name if no name was found in the user data
         if (pbtd.participant_name.name) {
           name = pbtd.participant_name.name;
         }
       }
     }
+
+    // ignore discovered participants without a name
     if (name.empty()) {
-      // ignore discovered participants without a name
-      node_names->data[i] = nullptr;
-      node_namespaces->data[i] = nullptr;
+      tmp_names_list.data[i] = nullptr;
+      tmp_namespaces_list.data[i] = nullptr;
       continue;
     }
 
-    node_names->data[i] = rcutils_strdup(name.c_str(), allocator);
-    if (!node_names->data[i]) {
+    tmp_names_list.data[named_nodes_num] = rcutils_strdup(name.c_str(), allocator);
+    if (!tmp_names_list.data[named_nodes_num]) {
       RMW_SET_ERROR_MSG("could not allocate memory for node name");
       goto fail;
     }
 
-    node_namespaces->data[i] = rcutils_strdup(namespace_.c_str(), allocator);
-    if (!node_namespaces->data[i]) {
+    ++named_nodes_num;
+    
+    if (namespace_.empty()) {
+      tmp_namespaces_list.data[i] = nullptr;
+      continue;
+    }
+    
+    tmp_namespaces_list.data[named_namespaces_num] = rcutils_strdup(namespace_.c_str(), allocator);
+    if (!tmp_namespaces_list.data[named_namespaces_num]) {
       RMW_SET_ERROR_MSG("could not allocate memory for node namespace");
       goto fail;
     }
+    
+    ++named_namespaces_num;
   }
 
+  // prepare actual output without empty names
+  // TODO: implement functionality that allocates just enough memory for the output buffer,
+  // so using additional buffer would be unnessecary 
+  rcutils_ret = rcutils_string_array_init(node_names, named_nodes_num, &allocator);
+  if (rcutils_ret != RCUTILS_RET_OK) {
+    RMW_SET_ERROR_MSG(rcutils_get_error_string().str);
+    return rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
+  }
+
+  for (auto i = 0; i < named_nodes_num; ++i) {
+    node_names->data[i] = rcutils_strdup(tmp_names_list.data[i], allocator);
+    tmp_names_list.data[i] = nullptr;
+  }
+
+  for (auto i = 0; i < named_nodes_num; ++i) {
+    std::cerr << tmp_names_list.data[i] << '\n';
+  }
+  rcutils_ret = rcutils_string_array_fini(&tmp_names_list);
+  if (rcutils_ret != RCUTILS_RET_OK) {
+    RMW_SET_ERROR_MSG(rcutils_get_error_string().str);
+    return rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
```

Comment:
Be careful returning too early, we are leaking memory since `tmp_namespaces_list` isn't being finalized.
I suggest refactoring so that we only return once at the end of the function. We can rename the `fail` label to `cleanup`, and whenever there is an error, we can set an error message, set a return code, and `goto cleanup`.

E.g. 

```c
rmw_ret_t final_ret = RMW_RET_OK;  // put this near the top after the initial error handling (e.g. L65)

// ...

if (rcutils_ret_ != RCUTILS_RET_OK) {
  RMW_SET_ERROR_MSG(rcutils_get_error_string().str);
  final_ret = rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
  goto cleanup;
}

// ...

cleanup:
// ... cleanup logic
return final_ret;
```

Reply:
Done